core/templates/dev/head/domain/state_card/StateCardObjectFactory.js

```diff
@@ -17,114 +17,189 @@
+        interactionId && INTERACTION_SPECS[interactionId].is_terminal);
+    };
+
+    StateCard.prototype.getInteractionInstructions = function() {
+      var interactionId = this.getInteractionId();
+      return (
+        interactionId ? INTERACTION_SPECS[interactionId].instructions : '');
+    };
+
+    StateCard.prototype.getInteractionCustomizationArgs = function() {
+      var interaction = this.getInteraction();
+      if (!interaction) {
+        return null;
+      }
+      return interaction.customizationArgs;
+    };
+
+    StateCard.prototype.isInteractionInline = function() {
+      var interactionId = this.getInteractionId();
+      return (
+        !interactionId ||
+        INTERACTION_SPECS[interactionId].display_mode ===
+          INTERACTION_DISPLAY_MODE_INLINE);
+    };
+
+    StateCard.prototype.getCurrentParams = function() {
+      return this._currentParams;
+    };
+
+    StateCard.prototype.getContentHtml = function() {
+      return this._contentHtml;
+    };
+
+    StateCard.prototype.getInteractionHtml = function() {
+      return this._interactionHtml;
+    };
+
+    StateCard.prototype.getOppiaResponse = function(index) {
+      return this._inputResponsePairs[index].oppiaResponse;
+    };
+
+    StateCard.prototype.getLeadsToConceptCard = function() {
+      return this._leadsToConceptCard;
+    };
+
+    StateCard.prototype.getInputResponsePairs = function() {
+      return this._inputResponsePairs;
+    };
+
+    StateCard.prototype.getLastInputResponsePair = function() {
+      if (this._inputResponsePairs.length === 0) {
+        return null;
+      }
+      return this._inputResponsePairs[this._inputResponsePairs.length - 1];
+    };
+
+    StateCard.prototype.getLastOppiaResponse = function() {
+      if (this.getLastInputResponsePair() === null) {
+        return null;
+      }
+      return this.getLastInputResponsePair().oppiaResponse;
+    };
+
+    StateCard.prototype.addInputResponsePair = function(inputResponsePair) {
+      this._inputResponsePairs.push(angular.copy(inputResponsePair));
+    };
+
+    StateCard.prototype.setDestStateName = function(destStateName) {
+      this._destStateName = destStateName;
+    };
+
+    StateCard.prototype.setOppiaResponse = function(index, response) {
+      this._inputResponsePairs[index].oppiaResponse = response;
+    };
+
+    StateCard.prototype.setLastOppiaResponse = function(response) {
+      this.setOppiaResponse(this._inputResponsePairs.length - 1, response);
+    };
+
+    StateCard.prototype.setInteractionHtml = function(interactionHtml) {
+      this._interactionHtml = interactionHtml;
+    };
+
+    StateCard.prototype.setLeadsToConceptCard = function(leadsToConceptCard) {
+      this._leadsToConceptCard = leadsToConceptCard;
+    };
+
+    StateCard.prototype.addRandomSuffixToContentHtml = function(randomSuffix) {
+      this._contentHtml = this._contentHtml + randomSuffix;
+    };
+
+    StateCard.prototype.addRandomSuffixToInteractionHtml =
```

Comment:
This seems like the wrong place for these two "addRandomSuffix" functions, it should be done when generating the state card instead right?

Reply:
I just named it that way. The suffix is calculated beforehand and passed in, so this just adds whatver is passed in to the respective field. Should I rename to addSuffix then?

Reply:
Hm, I don't get why this is something the state card needs to worry about, if it's just a transcript.
Why can't the exp engine service generate the random thing and come up with the final interaction HTML, before passing it to the card's constructor?

Reply:
Done, removed this.